src/sessions.ts

```diff
@@ -615,12 +612,14 @@ function attemptTransaction<TSchema>(
   }
 
   return promise.then(
-    () => {
+    value => {
       if (userExplicitlyEndedTransaction(session)) {
-        return;
+        return value;
       }
 
-      return attemptTransactionCommit(session, startTime, fn, options);
```

Comment:
Will it still be possible to get the legacy return value somehow (i.e. the `Document` returned by `session.commitTransaction()`)?

I think it would be an acceptable breaking change in mongosh if that goes away (right now our own `withTransaction` helper returns it), but it might be nice to avoid this. 

To avoid this, you could do some hackery to attach this secondary result to the returned Promise itself or return a tuple, e.g.
- `withTransaction<T>(fn: WithTransactionCallback<T>, options?: …): Promise<T> & { transactionResult: Promise<Document | undefined> }`
- `withTransaction<T>(fn: WithTransactionCallback<T>, options?: …): Promise<[callbackResult: T, transactionResult: Document|undefined]>`

(that would also maybe solve the problem mentioned above about distinguishing the callback returning `undefined` from an aborted transaction?)

Reply:
My only worry here is with this change `withTransaction` and `withSession` now diverge, and one of the goals of the ticket was to keep those two APIs the same for consistency. But I'm flexible.

Reply:
Maybe it could just be a separate function or item in `options` then that makes it return a tuple?

Reply:
We decided to proceed with returning the user's callback return value and not provide a way to get the old behavior